### PR TITLE
Device managment

### PR DESCRIPTION
**Pull request by k-chaney**
Leverage `torch.nn.Module` to handle easy transfers between devices and not have to worry about setting default tensor locations.

- `examples/mnist/conv_mnist.py` contains a GPU option that leverages this
- `BasePipeline` will now handle transfer to GPU if it is available by default (both the network and the batch)
- `bindsnet/pipeline/base_pipeline.py` contains a `recursive_to` function that handles transferring an arbitrary batch structure to the target device

### CODE REVIEW

File: bindsnet/network/network.py [hunk 1]
```diff
@@ -100,7 +100,8 @@ def __init__(
         self.layers = {}
         self.connections = {}
         self.monitors = {}
-        self.learning = learning
+        self.train(learning)
```
Comment (djsaunde) [inline bindsnet/network/network.py]: What's your reason for changing this?

File: bindsnet/network/nodes.py [hunk 2]
```diff
@@ -34,6 +35,7 @@ def __init__(
         :param tc_trace: Time constant of spike trace decay.
         :param trace_scale: Scaling factor for spike trace.
         :param sum_input: Whether to sum all inputs.
+        :param learning: Whether to be in learning or testing
```
Comment (djsaunde) [inline bindsnet/network/nodes.py]: Period at the end of this comment.

File: bindsnet/learning/__init__.py [hunk 3]
```diff
@@ -71,8 +71,8 @@ def update(self) -> None:
         if (
             self.connection.wmin != -np.inf or self.connection.wmax != np.inf
         ) and not isinstance(self, NoOp):
-            self.connection.w = torch.clamp(
-                self.connection.w, self.connection.wmin, self.connection.wmax
+            self.connection.w.clamp_(
```
Comment (djsaunde) [inline bindsnet/learning/__init__.py]: Nice!

File: bindsnet/network/__init__.py [hunk 4]
```diff
@@ -1,362 +1,4 @@
-import tempfile
-from typing import Dict, Optional
-
-import torch
-
-from .monitors import AbstractMonitor
-from .nodes import AbstractInput, Nodes
-from .topology import AbstractConnection
+from .network import Network, load
 from ..learning.reward import AbstractReward
```
Comment (djsaunde) [inline bindsnet/network/__init__.py]: Is this needed here?

File: bindsnet/network/topology.py [hunk 5]
```diff
@@ -363,15 +369,13 @@ def normalize(self) -> None:
         Normalize weights along the first axis according to total weight per target neuron.
         """
         if self.norm is not None:
-            shape = self.w.size()
-            self.w = self.w.view(
+            # get a view and modify in place
```
Comment (djsaunde) [inline bindsnet/network/topology.py]: Awesome!

File: bindsnet/network/network.py (hunk 1 above)
Comment (k-chaney) [reply bindsnet/network/network.py]: For backwards compatibility `self.learning` is kept throughout the code and `self.train()` (the method from `torch.nn.Module`) has been overloaded to include the writing of `self.learning` as these effectively do the same operation. We can use `network.train()` or `network.eval()` as pytorch uses them, however directly writing to `network.training` no longer works. Pytorch internally keeps track of `self.training`. We could switch completely without too much work.

Comment (djsaunde) [reply bindsnet/network/network.py]: Yeah, that's a good point.

Comment (Hananel-Hazan) [reply bindsnet/network/network.py]: What Pytorch internally keeps track of when using `self.training`?
How that it help us? beside been more compatible with PyTorch API?

File: bindsnet/network/topology.py [hunk 6]
```diff
@@ -157,21 +157,24 @@ def __init__(
         """
         super().__init__(source, target, nu, weight_decay, **kwargs)
 
-        self.w = kwargs.get("w", None)
-        if self.w is None:
+        w = kwargs.get("w", None)
```
Comment (Hananel-Hazan) [inline bindsnet/network/topology.py]: Is there a redone to this use of internal self.w and temporal w?

File: bindsnet/network/network.py (hunk 1 above)
Comment (k-chaney) [reply bindsnet/network/network.py]: https://github.com/pytorch/pytorch/blob/181792176d46b5e14b4d4c63292afb6670f5e27e/torch/nn/modules/module.py#L987

it is used for batchnorm and dropout mainly. When switching between training and testing people are used to calling `train()` and `eval()` on their models. It is just a function that sets the flag in every `torch.nn.Module` that is a child of the primary one that it was called from.

File: bindsnet/network/topology.py (hunk 6 above)
Comment (k-chaney) [reply bindsnet/network/topology.py]: I wanted to set it as a `Parameter` and you can't pass `None` into that, only `torch.Tensor`s. Thus for the initial logic of creating `self.w` this is done without assigning to `self.w` and instead creating a temporary `w` that doesn't have the same constraints.
